ERCS/erc-7677.md

```diff
@@ -0,0 +1,296 @@
+---
+eip: 7677
+title: Paymaster Web Service Capability
+description: A way for apps to communicate with smart wallets about paymaster web services
+author: Lukas Rosario (@lukasrosario), Dror Tirosh (@drortirosh), Wilson Cusack (@wilsoncusack), Kristof Gazso (@kristofgazso)
+discussions-to: https://ethereum-magicians.org/t/erc-7677-paymaster-web-service-capability/19530
+status: Draft
+type: Standards Track
+category: ERC
+created: 2024-04-03
+requires: 4337, 5792
+---
+
+## Abstract
+
+With [EIP-5792](./eip-5792.md), apps can communicate with wallets about advanced features via capabilities. This proposal defines a capability that allows apps to request that [ERC-4337](./eip-4337.md) wallets communicate with a specified paymaster web service. To support this, we also define a standardized API for paymaster web services.
+
+## Motivation
+
+App developers want to start sponsoring their users' transactions using paymasters. Paymasters are commonly used via web services. However, there is currently no way for apps to tell wallets to communicate with a specific paymaster web service. Similarly, there is no standard for how wallets should communicate with these services. We need both a way for apps to tell wallets to communicate with a specific paymaster web service and a communication standard for wallets to do so.
+
+## Specification
+
+One new [EIP-5792](./eip-5792.md) wallet capability is defined. We also define a standard interface for paymaster web services as a prerequisite.
+
+### Paymaster Web Service Interface
+
+We define two JSON-RPC methods to be implemented by paymaster web services.
+
+#### `pm_getPaymasterStubData`
+
+Returns stub values to be used in paymaster-related fields of an unsigned user operation for gas estimation. Accepts an unsigned user operation, entrypoint address, chain id, and a context object. Paymaster service providers can define fields that app developers should use in the context object. 
+
+##### `pm_getPaymasterStubData` RPC Specification
+
+```typescript
+// [userOp, entryPoint, chainId, context]
+type GetPaymasterStubDataParams = [
+  // Below is specific to Entrypoint v0.6 but this API can be used with other entrypoint versions too
+  {
+    sender: `0x${string}`;
+    nonce: `0x${string}`;
+    initCode: `0x${string}`;
+    callData: `0x${string}`;
+    callGasLimit: `0x${string}`;
+    verificationGasLimit: `0x${string}`;
+    preVerificationGas: `0x${string}`;
+    maxFeePerGas: `0x${string}`;
+    maxPriorityFeePerGas: `0x${string}`;
+  },
+  `0x${string}`,
```

Comment:
Why does the RPC specification jump straight to a TypeScript data shape without any explanation of semantics? I think it would be good to explain, for example, that the first of these hex strings is the address and the second the EIP155 chainId converted to hex (which would make sense given the examples), at the very least by normative reference over to [the exact section of 5792](https://eips.ethereum.org/EIPS/eip-5792#wallet_getcapabilities-rpc-specification), although now that I notice, it's also just a data shape without any semantics there as well.

Concretely, though, I would mention that in [one of the 5792 examples](https://eips.ethereum.org/EIPS/eip-5792#wallet_getcapabilities-example-return-value), there are TWO chainIds given with different capability objects in each, which would not match this TS definition, which models a flat structure with only one chainId (and only one capability!).

Reply:
above this it says `Accepts an unsigned user operation, entrypoint address, chain id, and a context object.` There is also a comment above the type `// [userOp, entryPoint, chainId, context]`. Added comments to each line to try to make it more clear though.

The specification here is related to a single `wallet_sendCalls` call, which is specific to a single chain id, where this type defines how a wallet would communicate with a paymaster service. The return value from `wallet_getCapabilities` is just expressing the available capabilities of the wallet per chain. I don't think I understand the relationship you're seeing between the 5792 example and what is written here.

Reply:
That works too! I think I just have a[n admittedly unpopular] preference for very explicit and verbose declarations of what each property means and what shape is allowed for its values, plus what to do if the wrong shape is passed.

My comment about the 5792 example was confused, apologies.  Here, `chainId` is just a top-level string (of which there can only be one), in the 5792 example I was pointing to a top-level chainId used as a key, whose value was an object full of capabilities. No action needed.

Reply:
gotcha. all good!